### PR TITLE
Feature: DNS API for IONOS cloud

### PR DESCRIPTION
**Pull request by zak905**
This change set extends the existing DNS API  script for IONOS dns_ionos.sh to support the [IONOS Cloud API](https://api.ionos.com/docs/dns/v1/). IONOS users have two alternatives for managing their DNS zones: the [IONOS Cloud API](https://api.ionos.com/docs/dns/v1/) and the [IONOS DNS API](https://developer.hosting.ionos.com/docs/dns) (which is supported by the current implementation). The objective is to allow users of the IONOS Cloud API to also use the script. 

As explained in the script header comment, the script decides which API to use based on the provided credentials. If `IONOS_PREFIX` and `IONOS_SECRET` are provided, the script will use the IONOS DNS API, else if `IONOS_TOKEN` is provided the script will use the IONOS Cloud API. 

Note: This contribution is done by IONOS DNS cloud team, the team responsible directly for the service.

### CODE REVIEW

**Comment by zak905** [comment]
Hi @Neilpang, 

can you please take a look when you have a chance ?

The DNS tests are passing successfully: https://github.com/ionos-cloud/acme.sh/actions/runs/8941008104

To make the tests pass, I had to make some changes to the acmetest project so that the TXT record content is randomized:  https://github.com/acmesh-official/acmetest/pull/26

For now, this PR uses my own fork, as you can see here: https://github.com/ionos-cloud/acme.sh/blob/add_ionos_cloud_script/.github/workflows/DNS.yml#L70

Also how can I update the documentation?

**Comment by Neilpang** [comment]
remove this change please:
image

**Comment by zak905** [comment]
Thanks for merging. Here the successful run for the latest triggered action: https://github.com/ionos-cloud/acme.sh/actions/runs/8969436076

**Comment by zak905** [comment]
@Neilpang I just wanted to draw your attention to this PR. Any thoughts ?

File: dnsapi/dns_ionos_cloud.sh [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  fi
+
+  return 1
+}
+
+_ionos_cloud_get_record() {
+  zone_id=$1
+  txtrecord=$2
+  # this is to transform the domain to lower case
+  fulldomain=$(printf "%s" "$3" | tr "[:upper:]" "[:lower:]")
```
Comment (Neilpang) [inline dnsapi/dns_ionos_cloud.sh:98]: please use `_lower_case()` function instead

File: dnsapi/dns_ionos_cloud.sh [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  # this is to transform record name to lower case
+  # IONOS Cloud API transforms all record names to lower case
+  _record_name=$(printf "%s" "$fulldomain" | cut -d . -f 1 | tr "[:upper:]" "[:lower:]")
```
Comment (Neilpang) [inline dnsapi/dns_ionos_cloud.sh:101]: _upper_case

File: dnsapi/dns_ionos_cloud.sh (hunk 1 above)
Comment (zak905) [reply dnsapi/dns_ionos_cloud.sh]: sure, done

File: dnsapi/dns_ionos_cloud.sh (hunk 2 above)
Comment (zak905) [reply dnsapi/dns_ionos_cloud.sh]: done

**Pull request merged**